modules/services/aerospace/default.nix

```diff
@@ -72,9 +72,37 @@ in
               description = "Default orientation for the root container.";
             };
             on-window-detected = lib.mkOption {
-              type = listOf str;
+              type = types.listOf (types.attrsOf types.anything);
```

Comment:
Is there any reason for `types.anything` instead of a proper sub module?

Reply:
I did try to come up with something more restricted (not much really) in #1168 but don't have much knowledge about aerospace and possible values of this toml table so we kind of backed up to accept everything. Just leaving it here in case we want to play with it a bit.
```
  callbackConf =
    with lib.types;
    oneOf [
      bool
      int
      float
      str
      path
      attrs
    ]
    // {
      description = "AeroSpace config callback (bool, int, float, str, path or attrs)";
    };
...
            on-window-detected = lib.mkOption {
              type = listOf (attrsOf (either callbackConf (listOf callbackConf)));
...
```

Reply:
> don't have much knowledge about aerospace and possible values of this toml table

Looking at the code & docs I believe: 
1.`run` command is an array of strings and it's required.
2. `if.app-id`, `if.workspace`, `if.window-title-regex-substring` and `if.app-name-regex-substring` are all optional and of type string
3. `if.during-aerospace-startup` is a boolean and optional 
4. `check-further-callbacks` is optional and of type boolean

I will leave it to @thuvasooriya if they want to add proper typing here. Otherwise, I can follow up with a PR once this is merged.

Reply:
Thankyou for the valuable PR @talhaHavadar, and @z0al for taking the time to go through the docs. This was my first Nix contribution ever so I didn't want to mess around too much.
My initial solution before I saw #1168 was
```
              type = listOf (attrsOf (oneOf [
                bool
                str
                attrs
              ]));
```
which was not rigid in terms of types but then I changed it according to the suggestion made by @Enzime in #1168.
I'll see if I can come up with anything better. Thanks again.

Reply:
https://github.com/thuvasooriya/nix-darwin/commit/49c1990eab619be2fc46528cc2d797316585edb3
let me know if this can be merged to this PR @Enzime @z0al
Tried to whip something up from the above doc description, not sure if it handles all the edge cases. but works with the test cases and my config

Reply:
Here is some information, so that some smart people can figure out what's wrong quickly. Hope it helps. I've tried some methods to make the fields with default values disappear only to fail.

current PR generated config - **works properly**
```toml
[[on-window-detected]]
run = "layout floating"

[on-window-detected.if]
app-name-regex-substring = "telegram|finder"

[[on-window-detected]]
run = "move-node-to-workspace 0"

[on-window-detected.if]
app-name-regex-substring = "ripes"
```

my shoddy implementation of proper types in https://github.com/thuvasooriya/nix-darwin/commit/49c1990eab619be2fc46528cc2d797316585edb3 - **doesn't work**

```toml
[[on-window-detected]]
check-further-callbacks = true
run = "move-node-to-workspace 0"

[on-window-detected.if]
app-id = ""
app-name-regex-substring = "ripes"
during-aerospace-startup = false
window-title-regex-substring = ""
workspace = ""
```

Reply:
If you change the type of `app-id` to be `types.nullOr types.str` and set the default value to `null`, it should omit it from the output

Reply:
> If you change the type of `app-id` to be `types.nullOr types.str` and set the default value to `null`, it should omit it from the output

Yeah that's one of the very first things i tried

```nix
                        app-id = lib.mkOption {
                          type = nullOr str;
                          default = null;
                          description = "The application ID to match (optional).";
                        };
```

```sh
error: builder for '/nix/store/6i98bgjlsmqsaqn5hz3id9p0ipchf3ln-aerospace.toml.drv' 
failed with exit code 1;
       last 1 log lines:
       > Error: Cannot convert data to TOML (null values are not supported)
```

Reply:
You can try to [filter the config](https://github.com/NixOS/nixpkgs/blob/0850b8d575c01f01d52084419fa5932046b29d7a/nixos/modules/services/security/kanidm.nix#L45C18-L45C68) of null values before passing the settings to`toml.generate`

Reply:
> You can try to [filter the config](https://github.com/NixOS/nixpkgs/blob/0850b8d575c01f01d52084419fa5932046b29d7a/nixos/modules/services/security/kanidm.nix#L45C18-L45C68) of null values before passing the settings to`toml.generate`

This is what i was looking for.. Thx!

Reply:
unfortunately replicating the filter is somewhat troublesome for me.
my guess is that the list of submodules is not playing nicely with filterAttrs.
I tried the following and still get the above cannot convert data to TOML, meaning the null is not filtered.

I tried the default functions, filterAttrs, filterAttrsRecursive
Then I tried to implement a custom function to iterate over each item in the submodule list. Which doesn't work either. Maybe there is some obvious mistake that I'm overlooking. 
Any help or direction would be appreciated. @Enzime @z0al
```nix
  filterOptionalFields = value:
    if lib.isAttrs value then
      let
        filteredAttrs = lib.mapAttrs (_: v: filterOptionalFields v) value;
        nonNullAttrs = lib.filterAttrs (_: v: v != null) value;
      in
        filteredAttrs // nonNullAttrs
    else if lib.isList value then
      lib.filter (x: x != null) (map filterOptionalFields value)
    else
      value;
  # filterOptionalFields = lib.filterAttrsRecursive (_: v: v != null);
  configFile = format.generate "aerospace.toml" (filterOptionalFields cfg.settings);
```

Reply:
Ah, I see. My bad. I didn't know that `filterAttrsRecursive` doesn't handle the case of lists.

Here is a custom implementation that works. I took the original `filterAttrsRecursive` code and added a condition for lists + changed the predicate signature to accept only values.

```nix
  # Code needs proper formatting
  filterAttrsRecursive =
    pred:
    set:
    lib.listToAttrs (
      lib.concatMap
        (name:
        let v = set.${name}; in
        if pred v then [
          (lib.nameValuePair name (
            if lib.isAttrs v then filterAttrsRecursive pred v
            else if lib.isList v then
              (
                map
                  (i:
                  if lib.isAttrs i then
                    filterAttrsRecursive pred i
                  else i)
                  (lib.filter pred v)
              )
            else v
          ))
        ] else [ ]
        )
        (lib.attrNames set)
    );
  filterNulls = filterAttrsRecursive (v: v != null);
  configFile = format.generate "aerospace.toml" (filterNulls cfg.settings);
```

Let's hope that Enzime has a trick up their sleeve so we don't have to keep this custom function :) 

P.S. Don't forget to change all optional attributes to default to null even things like `check-further-callbacks` as Aerospace behaves differently when it's set to `false` vs left out.

Reply:
cool! it works! thanks.
merged the changes, and yes every optional is set to null by default.